### PR TITLE
feat(navigation-guard) : #MEX-197 prevent user changing page by setting navigation-guard

### PR DESCRIPTION
**Pull request by sjollois**
## Describe your changes
Prevent user changing page by setting navigation-guard
link to entcore PR
## Checklist tests
Change project, try to reload without saving....
## Issue ticket number and link
[https://jira.support-ent.fr/browse/MEX-197](url)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/resources/public/template/main.html [hunk 1]
```diff
@@ -1,47 +1,48 @@
...
+                           ng-change="displayState.getDocument = false; getGGB();" file-format="'any'"></media-library>
         </div>
-        <div class="row">
-            <input ng-disabled="!name || name.trim() == ''" type="submit" class="button right-magnet" ng-click="saveDocument(name)" i18n-value="Confirmer"/>
-            <input type="button" class="cancel right-magnet horizontal-spacing-twice" ng-click="cancelSaveDocument()" i18n-value="Annuler"/>
+    </lightbox>
+    <lightbox show="displayState.name" on-close="displayState.name = false">
+        <div ng-if="displayState.name">
+            <div class="row">
+                <h1>Nom de votre projet Geogebra</h1>
```
Comment (BastienBenard) [inline src/main/resources/public/template/main.html]: Il faudrait un i18n.

**Review by BastienBenard — APPROVED** [review]

File: src/main/resources/public/template/main.html [hunk 2]
```diff
@@ -1,47 +1,58 @@
...
-    </div>
-    <div class="cell twelve-mobile right-magnet right-spacing center-pmb"
-         ng-if="!data.documentSelected && !data.documentSelected.metadata && !data.documentSelected.metadata.filename">
-        <span>[[fileName]]</span>
-    </div>
-    <div class="cell twelve-mobile right-magnet right-spacing center-pmb"
-        ng-if="data && data.documentSelected && data.documentSelected.metadata && data.documentSelected.metadata.filename">
-        <span>[[data.documentSelected.metadata.filename]]</span>
+<div guard-root="geogebra"  ng-click="updateBase64()">
+    <div class="row" style="z-index: 10">
```
Comment (FlorentMr) [inline src/main/resources/public/template/main.html]: Ca n'a pas l'air d'être du nouveau code, mais le style dans la div ça n'est pas très joli, peut être le remplacer si il y a du temps?

File: src/main/resources/public/ts/controllers/main.ts [hunk 3]
```diff
@@ -90,28 +105,42 @@ export const mainController = ng.controller('MainController', ['$timeout','$scop
...
+            const u8arr : Uint8Array = extractedFileBinary();
+            let file : File = new File([u8arr], name + ".ggb", {type: 'application/octet-stream'});
+            let doc : Document = new Document();
             workspace.v2.service.createDocument(file, doc, null,
                 {visibility: "protected", application: "media-library"}).then(async data => {
                 $scope.data.documentSelected = data;
                 http.post("geogebra");
                 toasts.confirm("Sauvegarde effectuée dans \"Documents ajoutés dans les applis\"");
                 $scope.displayState.name = false;
+                navigationGuardService.reset("geogebra");
```
Comment (FlorentMr) [inline src/main/resources/public/ts/controllers/main.ts]: magic string

File: src/main/resources/public/ts/controllers/main.ts [hunk 4]
```diff
@@ -90,28 +105,42 @@ export const mainController = ng.controller('MainController', ['$timeout','$scop
...
     }
 
-    $scope.saveDocument = (name: string) => {
+    $scope.saveDocument = (name: string) : void => {
         try {
-            const u8arr = extractedFileBinary();
-            let file = new File([u8arr], name + ".ggb", {type: 'application/octet-stream'});
-            let doc = new Document();
+            const u8arr : Uint8Array = extractedFileBinary();
+            let file : File = new File([u8arr], name + ".ggb", {type: 'application/octet-stream'});
```
Comment (FlorentMr) [inline src/main/resources/public/ts/controllers/main.ts]: magic string

File: src/main/resources/public/ts/controllers/main.ts [hunk 5]
```diff
@@ -90,28 +105,42 @@ export const mainController = ng.controller('MainController', ['$timeout','$scop
...
-    $scope.getGGBById = async () => {
+    const setGGBAplet = async (id: string) : Promise<void> => {
+        const file : AxiosResponse = await http.get(`workspace/document/base64/${id}`, {baseURL: '/'});
+        // @ts-ignore
+        ggbApplet.setBase64(file.data.base64File);
+        $scope.data.base64 = file.data.base64File;
+        const {fileName} = extractNameAndIdProject();
+        $scope.data.fileName = fileName;
+        await Utils.safeApply($scope);
+        navigationGuardService.reset("geogebra");
```
Comment (FlorentMr) [inline src/main/resources/public/ts/controllers/main.ts]: magic string

File: src/main/resources/public/ts/controllers/main.ts [hunk 6]
```diff
@@ -36,51 +37,65 @@ export const mainController = ng.controller('MainController', ['$timeout','$scop
...
         $scope.ggbApp = new GGBApplet(opts, true);
         $scope.ggbApp.inject('ggb-element');
         const int = setInterval(async function () {
             const app = $scope.ggbApp.getAppletObject();
             if (app && app.exists) {
                 clearInterval(int);
+                const hash : string = window.location.hash;
+                if(hash.length > 0){
+                    window.documentId = hash.substring(hash.indexOf("#/") + 2, hash.indexOf("?fileName="));
+                    window.fileName = hash.substring(hash.indexOf("?fileName=") + 10, hash.indexOf(".ggb") + 4)
```
Comment (FlorentMr) [inline src/main/resources/public/ts/controllers/main.ts]: magic string .ggb

File: src/main/resources/public/ts/controllers/main.ts [hunk 7]
```diff
@@ -120,54 +149,58 @@ export const mainController = ng.controller('MainController', ['$timeout','$scop
...
         while (n--) {
             u8arr[n] = byteCharacters.charCodeAt(n);
         }
         return u8arr;
     }
 
-    $scope.updateGGBFile = async () => {
+    function extractNameAndIdProject() : { fileName: string, id: string } {
+        let fileName : string = ($scope.data.documentSelected) ? $scope.data.documentSelected.metadata.filename : $scope.data.fileName;
+        if (!fileName.endsWith(".ggb")) fileName += ".ggb";
```
Comment (FlorentMr) [inline src/main/resources/public/ts/controllers/main.ts]: magic string

File: src/main/resources/public/ts/controllers/main.ts [hunk 8]
```diff
@@ -120,54 +149,58 @@ export const mainController = ng.controller('MainController', ['$timeout','$scop
...
+            const u8arr : Uint8Array = extractedFileBinary();
+            const {fileName, id} = extractNameAndIdProject();
+            let file : File = new File([u8arr], fileName, {type: 'application/octet-stream'});
+            let doc : Document = new Document();
             doc.name = fileName;
             doc._id = id;
             await workspace.v2.service.updateDocument(file, doc);
             http.post("geogebra");
             toasts.confirm("Sauvegarde effectuée");
+            navigationGuardService.reset("geogebra");
```
Comment (FlorentMr) [inline src/main/resources/public/ts/controllers/main.ts]: magic string

**Review by FlorentMr — APPROVED** [review]

**Review by JinDouang — APPROVED** [review]

**Pull request merged**